tests/unit_tests/test_analytics/sentry_relay.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import gzip
+import socket
+import threading
+import time
+from typing import Any
+
+import flask
+import requests
+from flask import request
+from sentry_sdk.envelope import Envelope
+
+
+class SentryResponse:
+    def __init__(
+        self,
+        message: str | None,
+        project_id: str,
+        public_key: str,
+        tags: None | dict[str, Any] = None,
+        is_error: bool = False,
+        stacktrace: None | dict[str, Any] = None,
+    ):
+        self.message = message
+        self.project_id = project_id
+        self.public_key = public_key
+        self.tags = tags
+        self.stacktrace = stacktrace
+        self.is_error = is_error
+
+    def __eq__(self, other):
+        return (
+            self.message == other.message
+            and self.project_id == other.project_id
+            and self.public_key == other.public_key
+            and self.tags == other.tags
+            and self.is_error == other.is_error
+        )
+
+
+class MetricRelayServer:
+    """
+    A mock Sentry Relay server that listens for local requests to sentry APIs.
+    These local requests are stored in a dictionary of event_id to SentryResponse.
+    """
+
+    events: dict = {}
+
+    def __init__(
+        self,
+    ) -> None:
+        self.is_running = False
+        self.port = self._get_free_port()
+        self.app = flask.Flask(__name__)
+        self.app.add_url_rule(
+            rule="/api/<project_id>/envelope/",
+            methods=["POST"],
+            view_func=self.sentry,
+        )
+        self.app.add_url_rule(
+            rule="/ping/",
+            methods=["GET"],
+            view_func=self.ping,
+        )
+        self.relay_url = f"http://127.0.0.1:{self.port}"
+
+        self._sentry_responses = []
```

Comment:
Is this used anywhere?

Reply:
It was only used when verifying the server setup, but we can manually construct the string there 😃